zerver/migrations/0260_missed_message_addresses_from_redis_to_db.py

```diff
@@ -0,0 +1,92 @@
+# -*- coding: utf-8 -*-
+from django.core.exceptions import ObjectDoesNotExist
+from django.db import migrations
+from django.db.backends.postgresql_psycopg2.schema import DatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+
+from zerver.lib.redis_utils import get_redis_client
+from zerver.lib.utils import generate_random_token
+
+def generate_missed_message_token() -> str:
+    return 'mm' + generate_random_token(32)
+
+def move_missed_message_addresses_to_database(apps: StateApps, schema_editor: DatabaseSchemaEditor) -> None:
+    redis_client = get_redis_client()
+    MissedMessageEmailAddress = apps.get_model('zerver', 'MissedMessageEmailAddress')
+    UserProfile = apps.get_model('zerver', 'UserProfile')
+    Message = apps.get_model('zerver', 'Message')
+    Recipient = apps.get_model('zerver', 'Recipient')
+    RECIPIENT_PERSONAL = 1
+    RECIPIENT_STREAM = 2
+
+    all_mm_keys = redis_client.keys('missed_message:*')
+    for key in all_mm_keys:
+        # Don't migrate mm addresses that have already been used.
+        if redis_client.hincrby(key, 'uses_left', -1) < 0:
+            redis_client.delete(key)
+            continue
+
+        result = redis_client.hmget(key, 'user_profile_id', 'recipient_id', 'subject')
+        if not all(val is not None for val in result):
+            # Missing data, skip this key
```

Comment:
Do we want to log an error in this case?  My understanding is this sort of case should never happen...

Reply:
I took this check from the original email_mirror code, but yeah, I'm not aware of any situations where this actually happens. I added a ``logging.error`` here.